storage/innobase/buf/buf0buf.cc

```diff
@@ -2804,83 +2814,78 @@ buf_page_get_low(
 #endif /* UNIV_DEBUG */
 	ut_ad(block->page.frame);
 
+	/* The state = block->page.state() may be stale at this point,
+	and in fact, at any point of time if we consider its
+	buffer-fix component. If the block is being read into the
+	buffer pool, it is possible that buf_page_t::read_complete()
+	will invoke buf_pool_t::corrupted_evict() and therefore
+	invalidate it (invoke buf_page_t::set_corrupt_id() and set the
+	state to FREED). Therefore, after acquiring the page latch we
+	must recheck the state. */
+
 	if (state >= buf_page_t::UNFIXED
 	    && allow_ibuf_merge
 	    && fil_page_get_type(block->page.frame) == FIL_PAGE_INDEX
 	    && page_is_leaf(block->page.frame)) {
 		block->page.lock.x_lock();
-		ut_ad(block->page.id() == page_id
-		      || (state >= buf_page_t::READ_FIX
-			  && state < buf_page_t::WRITE_FIX));
-
-#ifdef BTR_CUR_HASH_ADAPT
-		btr_search_drop_page_hash_index(block, true);
-#endif /* BTR_CUR_HASH_ADAPT */
-
-		dberr_t e;
-
-		if (UNIV_UNLIKELY(block->page.id() != page_id)) {
-page_id_mismatch:
-			state = block->page.state();
-			e = DB_CORRUPTION;
-ibuf_merge_corrupted:
-			if (err) {
-				*err = e;
-			}
-
-			if (block->page.id().is_corrupted()) {
-				buf_pool.corrupted_evict(&block->page, state);
-			}
-			return nullptr;
-		}
-
 		state = block->page.state();
 		ut_ad(state < buf_page_t::READ_FIX);
 
 		if (state >= buf_page_t::IBUF_EXIST
 		    && state < buf_page_t::REINIT) {
 			block->page.clear_ibuf_exist();
-			e = ibuf_merge_or_delete_for_page(block, page_id,
-							  block->zip_size());
-			if (UNIV_UNLIKELY(e != DB_SUCCESS)) {
-				goto ibuf_merge_corrupted;
+			if (dberr_t local_err =
+			    ibuf_merge_or_delete_for_page(block, page_id,
+							  block->zip_size())) {
+				if (err) {
+					*err = local_err;
+				}
+				goto release_and_ignore_block;
 			}
+		} else if (state < buf_page_t::UNFIXED) {
+release_and_ignore_block:
+			block->page.lock.x_unlock();
+			goto ignore_block;
 		}
 
-		if (rw_latch == RW_X_LATCH) {
-			goto get_latch_valid;
-		} else {
+#ifdef BTR_CUR_HASH_ADAPT
+		btr_search_drop_page_hash_index(block, true);
+#endif /* BTR_CUR_HASH_ADAPT */
+
+		switch (rw_latch) {
+		case RW_NO_LATCH:
+			block->page.lock.x_unlock();
+			break;
+		case RW_S_LATCH:
 			block->page.lock.x_unlock();
-			goto get_latch;
+			block->page.lock.s_lock();
+			break;
+		case RW_SX_LATCH:
+			block->page.lock.x_u_downgrade();
+			break;
+		default:
+			ut_ad(rw_latch == RW_X_LATCH);
 		}
+
+		mtr->memo_push(block, mtr_memo_type_t(rw_latch));
 	} else {
-get_latch:
 		switch (rw_latch) {
 		case RW_NO_LATCH:
+			/* Our caller is responsible for checking
+			block->page.is_freed(), that is, whether
+			buf_pool_t::corrupted_evict() had been called,
+			after acquiring the page latch, unless it is
+			known that the block was already successfully
+			loaded into the buffer pool. */
 			mtr->memo_push(block, MTR_MEMO_BUF_FIX);
 			return block;
```

Comment:
Sorry, this comment should not have been added. It turns out that we will already have acquired and released a page latch in case the block was read-fixed. So, at this point the block is guaranteed not to be corrupted, and the callers (mainly `rtr_search_to_nth_level()` and `rtr_pcur_getnext_from_path()`) should actually be safe.

Reply:
Removed in 3a570f75d4e1d5ad2e55189b60fe77437e71959c